### DISCUSSION TITLE
Source Generator instance lifetime in IDE

### DISCUSSION THREAD

**Discussion by VolodymyrBS** [dotnet/roslyn#50586] [Q&A] [answered]
I have a small question about Source Generator instance lifetime when it used by Visual Studio.

Does Visual Studio creat Source Generator just once and later just reuse it? or instance created each time generator should run?

I just curious if I could cache some data in instance field and use it on next generation. Or it's better to use other caching strategy?

thanks in advance!

**Comment by PathogenDavid** [comment, marked as answer]
From the [`ISourceGenerator`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.isourcegenerator?view=roslyn-dotnet) documentation:

> State should not be stored directly on the generator, as there is no guarantee that the same instance will be used on a subsequent generation pass.

The compiler makes no guarantee one way or another. It would be inadvisable to rely on the current behavior regardless of what it is.

**Reply by VolodymyrBS** [reply]
I totally missed it.
Thanks!
